vignettes/sir_models.Rmd

```diff
@@ -0,0 +1,680 @@
+
+[`new()`](https://mrc-ide.github.io/dust/reference/dust.html#method-new) takes the data needed to run the model i.e. a list with any parameters defined as `user` in the odin code above, the value of the initial step $t_0$, and the number of particles, each for now can simply be thought of as an independent stochastic realisation of the model, but in the next step will be used when inferring model parameters.
+
+Additional arguments include the number of threads to parallelise the particles over, and the seed for the random number generator. The seed must be an integer, and using the same seed will ensure reproducible results for all particles. For threads to take effect, `openmp` must be enabled in the compilation step, which you can check by running:
+
+```{r}
+gen_sir$public_methods$has_openmp()
+```
+
+To use this to directly create a new dust object with 10 particles, run using 4 threads:
+
+```{r}
+sir_model <- gen_sir$new(data = list(dt = 1, S_ini = 1000, I_ini = 10, beta = 0.2, gamma = 0.1),
+                         step = 1,
+                         n_particles = 10L,
+                         n_threads = 4L,
+                         seed = 1L)
+```
+
+Check the initial state is ten particles wide, with four states deep ($t$, $S$, $I$, $R$):
+
+```{r}
+sir_model$state()
+```
+
+Run the particles (repeats) forward 10 steps of length $dt$, followed by another 10 steps:
+
+```{r}
+sir_model$run(10)
+sir_model$run(20)
+```
+
+To change the parameters, the model does not need to be recompiled. We simply use the generator object to make a new model. Let's do this and increase the infection rate, and the population size. We will also use a smaller time step to calculate multiple transitions per unit time:
+
+```{r}
+dt = 0.25
+n_particles <- 10L
+sir_model_2 <- gen_sir$new(data = list(dt = 0.25, S_ini = 2000, I_ini = 10, beta = 0.4, gamma = 0.1),
+                           step = 1,
+                           n_particles = n_particles,
+                           n_threads = 4L,
+                           seed = 1L)
+sir_model_2$state()
+```
+
+Let's run this epidemic forward, and plot the trajectories:
+
+```{r fig.height=5, fig.width=7}
+n_steps <- 200
+x <- array(NA, dim = c(sir_model_2$info()$len, n_particles, n_steps))
+
+for (t in seq_len(n_steps)) {
+  x[ , , t] <- sir_model_2$run(t)
+}
+time <- x[1, 1, ]
+x <- x[-1, , ]
+
+par(mar = c(4.1, 5.1, 0.5, 0.5), las = 1)
+matplot(time, t(x[1, , ]), type = "l",
+         xlab = "Time", ylab = "Number of individuals",
+         col = "#8c8cd9", lty = 1, ylim = range(x))
+matlines(time, t(x[2, , ]), col = "#cc0044", lty = 1)
+matlines(time, t(x[3, , ]), col = "#999966", lty = 1)
+legend("left", lwd = 1, col = c("#8c8cd9", "#cc0044", "#999966"), legend = c("S", "I", "R"), bty = "n")
+```
```

Comment:
Consider something above like

```
cols <- c(S = "#8c8cd9", I = "#cc0044", R = "#999966")
```

so that you can do

```
matplot(..., col = cols[["S"]])
matlines(..., col = cols[["I"]])
legend(..., col = cols, legend = name(cols))
```